packages/grid/src/components/Container.tsx

```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import cn from 'classnames';
+
+import { ContainerProps } from './Container.types';
+
+import styles from './Container.module.css';
+
+function Container<E extends React.ElementType = 'div'>({
```

Comment:
Maybe C could represent `container` (also in `Container.types.ts`)?. 

`ElementType = 'div'` could possibly be moved to the types file?

Another possible way to simplify a bit and transfer the typing logic to just one place, something like this could be done:

```
export interface PropsBasedOnComponent<T> {
  <C extends React.ElementType = 'div'>(
    props: {
      /**
       * The component used for the root node.
       * Either a string to use a HTML element or a component.
       */
      component?: C;
    } & T &
      Omit<React.ComponentPropsWithoutRef<C>, keyof T>,
  ): JSX.Element | null;
}

const Container: PropsBasedOnComponent<ContainerProps> = ({
  component,
  ...rest
}) => {
```